### ISSUE TITLE
Three Ways to Improve Cordova/Meteor on iOS

### ISSUE THREAD

**Issue by msavin** [meteor/meteor#6718]
First up, huge thanks to MDG to improving Cordova and making it professional grade. I've been enthusiastically using it the past week or so, and have some suggestions to improve the developer experience around it. 

**Opportunity #1: Have Cordova connect over localhost:3000**

The new version connects over the existing network (ie. wi-fi network) and uses that to communicate with Meteor. This is problematic if you are not always connected to wi-fi, or switching networks (as the IP address changes)

The solution advised was to run `meteor run ios-device --mobile-server localhost:3000` but it came with another catch.When you run the application this way you 1) have to rebuild Cordova each time you run it, and 2) it runs the application in production mode.

To get around production mode, I'd terminate the server, and start it it again with just `meteor` on port 3000. This causes hot code push to break, while Cordova gives the following error in the console:

`Error: Error parsing asset manifest: Error Domain=NSCocoaErrorDomain Code=3840 "JSON text did not start with array or object and option to allow fragments not set." UserInfo={NSDebugDescription=JSON text did not start with array or object and option to allow fragments not set.}
 - column: 30
 - line: 36
 - message: "Error parsing asset manifest: Error Domain=NSCocoaErrorDomain Code=3840 \"JSON text did not start with array or object and option to allow fragments not set.\" UserInfo={NSDebugDescription=JSON text did not start with array or object and option to allow fragments not set.}"
 - sourceURL: "http://localhost:12272/plugins/cordova-plugin-meteor-webapp/www/webapp_local_server.js"
 - stack: "http://localhost:12272/plugins/cordova-plugin-meteor-webapp/www/webapp_local_server.js:36:30↵callbackFromNative@http://localhost:12272/cordova.js:293:63↵http://localhost:12272/plugins/cordova-plugin-wkwebview-engine/src/www/ios/ios-wkwebview-exec.js:136:32"`

**Opportunity #2: Make Meteor-Cordova package transparent**

If you go to ./meteor/cordova-plugins, you'll notice that it does not mention the Cordova plug-in that Meteor automatically bundles. Mentioning it there would make your app more transparent, and give you the ability to tell Meteor to look for the plugin on the computer than download it via git.

Why is it important to run Cordova plug-ins locally? Well, each time you run your Cordova app, it has to download all the packages and then build it. This takes a lot of time even on a good connection, and it's not possible when you are not connected to the internet. 

I've been running my plug-ins locally with this approach:

`in ./meteor/cordova-plugins
com.local.statusbar@file://../aqua-cordova-plugins/statusbar`

It saves me a lot of time, lets me know what's running in Cordova, and even let's me make a tweak to the plug-in if needed. It would be nice if I can attach the `@file://` to the Meteor plug-in and run it, debug it, etc, easily.

If the developer removes the package, then Meteor should warn them and automatically add the package to the top of the cordova-plugins list.

**Opportunity #3: Run iPhone 6S as the default device**

The iPhone 6 has a reasonable 2x resolution that fits neatly into any laptops screen size, and perfectly on Retina displays. In contrast, the iPhone 6+ has a huge 3x resolution that is scaled down to 1080p on the physical device. If you emulate the 6+ on the computer, you get a huge device, and even if you scale it down to 25%, it's rather big and looks distorted.

The iPhone 6/6S is the "right" choice because it has a reasonable resolution, fits seamlessly on our computers, and is the most post popular product in the series. While you can modify the setting by having Meteor open Xcode, its just easier (and I believe, right) to have it as the default.

The end goal here is to be able to run `meteor run ios`, or just `meteor`, without having to worry about configuration, slow rebuilds that require internet, or even the need to be a network.

I have yet to get to Android with this. I had some issues on the previous version but decided to wait for the new update. What I find lacking in this department is documentation on how to push updated versions for Android. I've read there are some challenges around versioning, but I haven't looked deep enough into it yet.

Compared to the previous Cordova version, this has made development much harder. Before, I'd be able to run meteor,  open up iOS simulator, and get to work. Currently, I have to rebuild whenever I switch networks.

**Comment by martijnwalraven** [comment]
I think it would be easier to respond if these were separate issues. But let me respond to your first point first.

I'm not sure what you mean by `meteor run ios-device --mobile-server localhost:3000` running your app in production mode, and I'm not seeing how `meteor run` without mobile targets would change this. 

The error you're seeing is because without a mobile target, `meteor run` will no longer build the Cordova bundle (see #6417) and `/__cordova/manifest.json` will then return the default index page. It may make sense to improve the error message, but the behavior itself is intended.

'meteor run ios-device` has always required a network IP, because by opening Xcode you could be running your app on a remote device, and that will never work with `localhost`. The only difference with previous versions is that `meteor run ios` will now also use a network IP if one can be detected. The reason for this is that existing apps running on remote devices would otherwise be unable to connect after the first hot code push (because this would change the `ROOT_URL` to `localhost`).

So it's not clear to me what you are proposing. If you want the previous behavior, you can still pass in ` --mobile-server localhost:3000`.

**Comment by martijnwalraven** [comment]
On to point 2: 

There is a difference between plugins installed as dependencies of packages, and plugins installed directly (with `meteor add cordova:...`). Only the latter are listed in `./meteor/cordova-plugins`. Other plugins are considered indirect dependencies, and the way to change these is by changing the `package.js` for the package that adds the dependency.

Direct dependencies will always override indirect dependencies however. So you could simply override `cordova-plugin-statusbar` with your local version. Note that this will make builds slower however, because we will remove and add back all Cordova plugins on every build if a local plugin has been added. This is to avoid problems with Cordova's handling of dependencies between plugins.

**Comment by martijnwalraven** [comment]
Point 3: 

The default iOS simulator device is set by Cordova. I'm not sure what the reasoning behind the current choice is, but if we want we could override it.

**Comment by msavin** [comment]
Point 1: 

Problem A: if I run `meteor run ios`, it'll use my computers IP as the mobile server. 
 - Problem: If I am off the network, or it changes, I have to rebuild.
 - Meteor.absoluteUrl returns `http://192.168.1.101:3000/` or whatever my wifi ip is.

Problem B: When I run it `meteor run ios --mobile-server localhost:3000`
- Meteor.absoluteUrl returns `http://localhost:3000/` - 
- my debugPackages no longer show up
  - When my debugOnly packages dissapeared, I assumed that the command runs it in production mode. However, now I think there could be another issue. It seems related to this: 
https://github.com/meteor/meteor/issues/5589
 - when my debugOnly packages dissapears, I assumed that I need to re-run the server in development mode, then wait for HCP. I tried and tried, but HCP does not seem to work.
 - this is where it gets crazy: I've seen it with my own eyes, multiple times, but I am not able to reproduce it right now

Problem C: When I run it with `meteor run ios-device --mobile-server localhost:3000` 
  - in short, HCP breaks
  - in detail: to test for reliability, I would turn off the wi-fi, close the simulator, and shut down the server. Then, I would spin up the server, open iOS simulator, and try to get back into development. What I see is that my app connects to the server on localhost:3000, but hot code push no longer works. HCP in browsers work fast.
 - not using reload-on-resume, tried doing manual reloads in Cordova.

With the previous Cordova, I would build it once, and then always run it off iOS simulator. Whenever my app would change, I would depend on HCP to test those changes, since Cordova config did not change frequently much. It would work independent of my network, etc.

I've tested this a few times now to be sure. My app, environment and Meteor build are all standard. I hope this is helpful to you in diagnosing bugs, and am happy to help however I could. I hope we can get the development experience to where it was with 1.2 with the prowess of the new integration.

**Comment by msavin** [comment]
I think that was the biggest issue I am having, the other two: 

#2 - Could we cache the Cordova plug-ins that Meteor packages depend on? I think this already happens with packages. However - I'm not 100% sure which package brings in the Meteor-Cordova package, I always thought it happens over something in the .meteor/platforms file. 

#3 - I think an override would be a good idea, I'm strongly advocating for this. It'd be nice to hear what other's think.

**Comment by martijnwalraven** [comment]
Have you been able to find a reproduction of #5589? I'm not sure what could be going on there, but it doesn't seem related to the change in IP autodetection.

What seems to be going on in problem C is what I described in my previous response: if you restart the server by running `meteor` without a mobile target (so without either `ios` or `ios-device`) the Cordova bundle won't be built and hot code push for Cordova won't work.

**Comment by martijnwalraven** [comment]
About point 2: 
What do you mean by caching the Cordova plugins Meteor packages depend on? Plugins that are installed from npm are already cached.

**Comment by msavin** [comment]
#2: So, whenever I run Meteor with a mobile target, it downloads the Cordova plug-ins again. 

![screen shot 2016-04-07 at 11 18 39 am](https://cloud.githubusercontent.com/assets/2031235/14340504/a0385752-fcb2-11e5-839f-b29c17df3b6d.png)

It usually takes me a good minute to download the Meteor/Cordova package. The rest of my Cordova plug-ins, I run locally to cut down on the time. 

This is problematic UX because one cannot start Cordova, and therefore use Cordova, unless they have an internet connection.

**Comment by msavin** [comment]
On the reproduction of #5589: I have no idea - all I can report is when I run Meteor with a mobile server target it sometimes excludes my debugOnly packages.

`@martijnwalraven : What seems to be going on in problem C is what I described in my previous response: if you restart the server by running meteor without a mobile target (so without either ios or ios-device) the Cordova bundle won't be built and hot code push for Cordova won't work.`

On this subject, can we get Cordova working like the way it used to, or was this tradeoff intentional?

Before, I would be able to spin up iOS simulator and a server, they'd connect (provided they are both on the same port), and it would work whether its online or off. It was a much easier and faster experience, and it worked regardless of my network connection.

**Comment by martijnwalraven** [comment]
Meteor doesn't control caching of Cordova plugins, but Cordova should use an npm cache. It won't cache plugins added from Git however. Adding plugins from the local filesystem complicates this further, because it means we will remove and add plugins on each build, and that will trigger redownloads of Git plugins. If you only use plugins from npm, building your project shouldn't have to reinstall them.

The decision not to build for Cordova when not running on mobile targets was intentional, see https://github.com/meteor/meteor/issues/6417. But `meteor run ios --server localhost:3000` should give you exactly the same experience as before; I still don't understand what changed for you.

**Comment by msavin** [comment]
It seems there was a decision made to:
1) require you to run Cordova as part of running Meteor 
2) require you to pass in localhost:3000 manually

As a Meteor developer, this created a few issues for me:
- I have to be connected to the internet in order to start my Meteor app
- I need to rebuild the entire Cordova thing every time I manually restart the server, which means I have to re-download all plug-ins
- I was not aware of these changes, did not find information about them anywhere

Meteor developers jumping into this new implementation with the "old" habits will get confused when things do not work as expected. I am hoping we can go back to the "it just works" way of the previous versions.

I know this is opinionated stuff - but my take is that skipping the Cordova rebuilds as per issue #6417 is not good because it departs from the way the app works in production, and the way we have been working with it. If there's a server running, all the clients should operate as you would expect. Maybe there should be an option to `meteor run web-only` if someone feels slowed down by the Cordova rebuilds.

**Comment by martijnwalraven** [comment]
The point is: you still don't have to be connected to the internet to start your Meteor app! I just tried this out on a new mobile app, and `meteor run ios` while you're not connected to the internet just works. 

What changed is that `meteor run ios` will now autodetect your IP address if you ARE connected to an external network, and will use that instead of `localhost`. This change was made because the previous behavior was confusing: if you had the app running on a device hot code pushing from a Meteor server started with `meteor run ios`, it would change its `ROOT_URL` to `localhost` and would no longer be able to connect. But if you want the previous behavior, `meteor run ios --server localhost:3000` should work exactly the same.

The only situation where this won't work is if you have Cordova plugins installed from the local filesystem. This will indeed force a reinstallation of all plugins on every build (as it has always done), and ONLY for plugins installed from Git, that will involve redownloading, which obviously won't work if you are offline.

A separate issue is whether skipping the Cordova builds when you're not running with a mobile target is a good idea. A case was made in #6417 for this, and it sounds reasonable to me. Your use case is different, but also fairly specific. If you feel we should reevaluate this change, let's continue the conversation there and let other people chime in.

**Comment by martijnwalraven** [comment]
Closing this as part of issue cleanup. Please open separate issues with feature requests if you still feel this is something you'd like to discuss.
